In [1]:
import pandas as pd
import numpy as np

In [2]:
raw_data = pd.read_csv('data.csv')
raw_data.head()

register_index  post_id comment_id                author  \
0     145k12dpost  145k12d        NaN               SauloJr   
1  145hyeyjnlh1vx  145hyey    jnlh1vx  Illustrious_Risk3732   
2  1453u2ojnjtk7f  1453u2o    jnjtk7f       ProdigiousPlays   
3  1453u2ojnjtv9u  1453u2o    jnjtv9u                YJSubs   
4  1453u2ojnjusm9  1453u2o    jnjusm9           jamnewton22   

              datetime                                              title  \
0  2023-06-09 23:17:15  Reddit CEO Spez answered a grand total of 14 q...   
1  2023-06-09 23:42:43                                                NaN   
2  2023-06-09 16:47:51                                                NaN   
3  2023-06-09 16:49:52                                                NaN   
4  2023-06-09 16:55:57                                                NaN   

                                                 url  score  comments  \
0  https://www.reddit.com/r/technology/comments/1...    1.0       0.0   
1                                                NaN    1.0       NaN   
2                                                NaN    1.0       NaN   
3                                                NaN    1.0       NaN   
4                                                NaN    1.0       NaN   

                                                text  author_post_karma  \
0  Reddit CEO Spez answered a grand total of 14 q...            44092.0   
1  He even gives him self Awards to make sure he ...            17109.0   
2  RIF already announced it s shutting down at th...           118844.0   
3              I wonder which sub left on that day              50085.0   
4                  Oh I m sure this will go well lol           171705.0   

            tag  
0  Social Media  
1  Social Media  
2  Social Media  
3  Social Media  
4  Social Media

In [3]:
print(raw_data.shape)

(86916, 12)


In [4]:
raw_data.loc[raw_data['comment_id'].isna()]

register_index  post_id comment_id             author  \
0        145k12dpost  145k12d        NaN            SauloJr   
125      145g5xfpost  145g5xf        NaN     lily__brunette   
348      145f8kvpost  145f8kv        NaN  thebelsnickle1991   
431      1447ctspost  1447cts        NaN            ionedew   
455      145xmyxpost  145xmyx        NaN          dasun0218   
...              ...      ...        ...                ...   
86737    149fnx5post  149fnx5        NaN              hawlc   
86740    149fh87post  149fh87        NaN         RebootJobs   
86742    149eob7post  149eob7        NaN          OneOk2189   
86781    149egcvpost  149egcv        NaN         spsheridan   
86793    149drwspost  149drws        NaN    giuliomagnifico   

                  datetime                                              title  \
0      2023-06-09 23:17:15  Reddit CEO Spez answered a grand total of 14 q...   
125    2023-06-09 20:40:07  Reddit CEO Steve Huffman defends API changes i...   
348    2023-06-09 20:03:31  Netflix's password crackdown leads to massive ...   
431    2023-06-08 12:01:58  Sam Altman says he worries making ChatGPT was ...   
455    2023-06-10 11:17:50  16 Game-Changing AI Tools: Unleashing the Futu...   
...                    ...                                                ...   
86737  2023-06-14 18:46:14  Google risks forced breakup of ad business as ...   
86740  2023-06-14 18:38:56  Government Releases Unusually Frank Report on ...   
86742  2023-06-14 18:05:54  Paul Graham says remote work ‘does work initia...   
86781  2023-06-14 17:57:09  A team of physicists recently announced that t...   
86793  2023-06-14 17:30:04  Iran's 'quantum processor' turned out to be a ...   

                                                     url   score  comments  \
0      https://www.reddit.com/r/technology/comments/1...     1.0       0.0   
125    https://www.reddit.com/r/technology/comments/1...   859.0     232.0   
348    https://www.reddit.com/r/technology/comments/1...     0.0      14.0   
431    https://www.reddit.com/r/technology/comments/1...     3.0      10.0   
455    https://www.reddit.com/r/technology/comments/1...     3.0       0.0   
...                                                  ...     ...       ...   
86737  https://www.reddit.com/r/technology/comments/1...    20.0       2.0   
86740  https://www.reddit.com/r/technology/comments/1...     1.0       1.0   
86742  https://www.reddit.com/r/technology/comments/1...     0.0      39.0   
86781  https://www.reddit.com/r/technology/comments/1...    56.0      10.0   
86793  https://www.reddit.com/r/technology/comments/1...  1561.0     129.0   

                                                    text  author_post_karma  \
0      Reddit CEO Spez answered a grand total of 14 q...            44092.0   
125    Reddit CEO Steve Huffman defends API changes i...              542.0   
348    Netflix s password crackdown leads to massive ...           176771.0   
431    Sam Altman says he worries making ChatGPT was ...              466.0   
455    16 Game Changing AI Tools  Unleashing the Futu...              477.0   
...                                                  ...                ...   
86737  Google risks forced breakup of ad business as ...             3434.0   
86740  Government Releases Unusually Frank Report on ...            21171.0   
86742  Paul Graham says remote work  does work initia...             4627.0   
86781  A team of physicists recently announced that t...            17704.0   
86793  Iran s  quantum processor  turned out to be a ...            94361.0   

                           tag  
0                 Social Media  
125               Social Media  
348                   Business  
431    Artificial Intelligence  
455    Artificial Intelligence  
...                        ...  
86737                 Business  
86740                  Privacy  
86742                  Society  
86781       Nanotech/Materials  
86793                 Hardw

In [5]:
raw_data.loc[raw_data['text'].isna()]

Empty DataFrame
Columns: [register_index, post_id, comment_id, author, datetime, title, url, score, comments, text, author_post_karma, tag]
Index: []

In [6]:
# of posts and comments inside it
raw_data.groupby('post_id').count()['comment_id'].sort_values()

post_id
14a6xf2      0
1453sm0      0
1459yy6      0
145e9we      0
145k12d      0
          ... 
14bs8v6    790
14h1tdg    838
14g6hin    843
14ag85h    866
14b11yr    905
Name: comment_id, Length: 975, dtype: int64

In [7]:
raw_data.loc[raw_data['post_id']=='14b11yr']

register_index  post_id comment_id          author  \
7148   14b11yrjoe9sxp  14b11yr    joe9sxp  TorontoCorsair   
7149   14b11yrjoelsez  14b11yr    joelsez       olearyboy   
7150   14b11yrjodpccr  14b11yr    jodpccr     HombreMan24   
7151   14b11yrjodn1rb  14b11yr    jodn1rb         franker   
7152   14b11yrjoeefjx  14b11yr    joeefjx     Danteynero9   
...               ...      ...        ...             ...   
77314  14b11yrjofudap  14b11yr    jofudap           W3NTZ   
77315  14b11yrjog5cqk  14b11yr    jog5cqk       SomniumOv   
77316  14b11yrjofityh  14b11yr    jofityh      Baliverbes   
77317  14b11yrjof52d5  14b11yr    jof52d5     Good_ApoIIo   
77318  14b11yrjog22xs  14b11yr    jog22xs     Good_ApoIIo   

                  datetime title  url  score  comments  \
7148   2023-06-16 20:48:31   NaN  NaN    3.0       NaN   
7149   2023-06-16 22:30:07   NaN  NaN    3.0       NaN   
7150   2023-06-16 18:23:20   NaN  NaN    7.0       NaN   
7151   2023-06-16 18:07:00   NaN  NaN    5.0       NaN   
7152   2023-06-16 21:22:52   NaN  NaN    4.0       NaN   
...                    ...   ...  ...    ...       ...   
77314  2023-06-17 05:28:54   NaN  NaN    6.0       NaN   
77315  2023-06-17 07:59:15   NaN  NaN    6.0       NaN   
77316  2023-06-17 03:21:04   NaN  NaN    5.0       NaN   
77317  2023-06-17 01:13:02   NaN  NaN   56.0       NaN   
77318  2023-06-17 07:11:48   NaN  NaN    4.0       NaN   

                                                    text  author_post_karma  \
7148   First you get the higher mod priviledges  Then...             5564.0   
7149                                      r fuckyoukaren            23873.0   
7150   Can someone enlighten me in why a mod actually...             2473.0   
7151   DON T MAKE US HAVE TO STOP LETTING YOU WORK FO...            76064.0   
7152     What are you going to do  take away my salary              17317.0   
...                                                  ...                ...   
77314  My thought is if people come to reddit for hel...           293576.0   
77315    I heard on a podcast that they re undeleting...           171592.0   
77316  thankfully there s  r eyebleach  ah  shit  the...            44121.0   
77317  No don t mistake my meaning  They re still her...           251925.0   
77318  Gross  well there s our timeline it was only b...           251925.0   

                tag  
7148   Social Media  
7149   Social Media  
7150   Social Media  
7151   Social Media  
7152   Social Media  
...             ...  
77314  Social Media  
77315  Social Media  
77316  Social Media  
77317  Social Media  
77318  Social Media  

[906 rows x 12 columns]

In [8]:
# Example of a post with its commments
print('Number of comments',raw_data.loc[raw_data['post_id']=='14b11yr'].shape[0]-1)
print('Number of unique authors',raw_data.loc[raw_data['post_id']=='14b11yr']['author'].nunique())

Number of comments 905
Number of unique authors 654


## Find out whether a comment can be commented.

In [9]:
sum(raw_data.groupby('comment_id').count()['comments'].values)

0

## Filter posts without comments

In [10]:
raw_data.loc[(raw_data['comment_id'].isna()) & (raw_data['comments']>0)]

register_index  post_id comment_id             author  \
125      145g5xfpost  145g5xf        NaN     lily__brunette   
348      145f8kvpost  145f8kv        NaN  thebelsnickle1991   
431      1447ctspost  1447cts        NaN            ionedew   
456      145vqmzpost  145vqmz        NaN             Majnum   
509      1466uytpost  1466uyt        NaN             777fer   
...              ...      ...        ...                ...   
86737    149fnx5post  149fnx5        NaN              hawlc   
86740    149fh87post  149fh87        NaN         RebootJobs   
86742    149eob7post  149eob7        NaN          OneOk2189   
86781    149egcvpost  149egcv        NaN         spsheridan   
86793    149drwspost  149drws        NaN    giuliomagnifico   

                  datetime                                              title  \
125    2023-06-09 20:40:07  Reddit CEO Steve Huffman defends API changes i...   
348    2023-06-09 20:03:31  Netflix's password crackdown leads to massive ...   
431    2023-06-08 12:01:58  Sam Altman says he worries making ChatGPT was ...   
456    2023-06-10 09:31:26  Recent Mars Research Reveals New Discoveries a...   
509    2023-06-10 18:05:28  There's been a whopping 736 Tesla crashes invo...   
...                    ...                                                ...   
86737  2023-06-14 18:46:14  Google risks forced breakup of ad business as ...   
86740  2023-06-14 18:38:56  Government Releases Unusually Frank Report on ...   
86742  2023-06-14 18:05:54  Paul Graham says remote work ‘does work initia...   
86781  2023-06-14 17:57:09  A team of physicists recently announced that t...   
86793  2023-06-14 17:30:04  Iran's 'quantum processor' turned out to be a ...   

                                                     url   score  comments  \
125    https://www.reddit.com/r/technology/comments/1...   859.0     232.0   
348    https://www.reddit.com/r/technology/comments/1...     0.0      14.0   
431    https://www.reddit.com/r/technology/comments/1...     3.0      10.0   
456    https://www.reddit.com/r/technology/comments/1...    15.0       2.0   
509    https://www.reddit.com/r/technology/comments/1...    26.0      19.0   
...                                                  ...     ...       ...   
86737  https://www.reddit.com/r/technology/comments/1...    20.0       2.0   
86740  https://www.reddit.com/r/technology/comments/1...     1.0       1.0   
86742  https://www.reddit.com/r/technology/comments/1...     0.0      39.0   
86781  https://www.reddit.com/r/technology/comments/1...    56.0      10.0   
86793  https://www.reddit.com/r/technology/comments/1...  1561.0     129.0   

                                                    text  author_post_karma  \
125    Reddit CEO Steve Huffman defends API changes i...              542.0   
348    Netflix s password crackdown leads to massive ...           176771.0   
431    Sam Altman says he worries making ChatGPT was ...              466.0   
456    Recent Mars Research Reveals New Discoveries a...            17593.0   
509    There s been a whopping 736 Tesla crashes invo...              105.0   
...                                                  ...                ...   
86737  Google risks forced breakup of ad business as ...             3434.0   
86740  Government Releases Unusually Frank Report on ...            21171.0   
86742  Paul Graham says remote work  does work initia...             4627.0   
86781  A team of physicists recently announced that t...            17704.0   
86793  Iran s  quantum processor  turned out to be a ...            94361.0   

                           tag  
125               Social Media  
348                   Business  
431    Artificial Intelligence  
456                      Space  
509             Transportation  
...                        ...  
86737                 Business  
86740                  Privacy  
86742                  Society  
86781       Nanotech/Materials  
86793                 Hardw

The total number of posts is 975. Total number of posts with at least 1 comment is 909. There are 66 posts without comments. We can exclude them.

In [11]:
posts = raw_data.loc[(raw_data['comment_id'].isna()) & (raw_data['comments']>0)]
posts.shape

(909, 12)

In [12]:
posts.isna().sum()

register_index         0
post_id                0
comment_id           909
author                 0
datetime               0
title                  0
url                    0
score                  0
comments               0
text                   0
author_post_karma     12
tag                    0
dtype: int64

In [13]:
comments = raw_data.loc[(raw_data['post_id'].isin(set(posts['post_id'])))&(~raw_data['comment_id'].isna())]
print(comments.shape)
comments.head()

(85941, 12)


register_index  post_id comment_id                author  \
1  145hyeyjnlh1vx  145hyey    jnlh1vx  Illustrious_Risk3732   
2  1453u2ojnjtk7f  1453u2o    jnjtk7f       ProdigiousPlays   
3  1453u2ojnjtv9u  1453u2o    jnjtv9u                YJSubs   
4  1453u2ojnjusm9  1453u2o    jnjusm9           jamnewton22   
5  1453u2ojnjvwlv  1453u2o    jnjvwlv            TriLink710   

              datetime title  url  score  comments  \
1  2023-06-09 23:42:43   NaN  NaN    1.0       NaN   
2  2023-06-09 16:47:51   NaN  NaN    1.0       NaN   
3  2023-06-09 16:49:52   NaN  NaN    1.0       NaN   
4  2023-06-09 16:55:57   NaN  NaN    1.0       NaN   
5  2023-06-09 17:03:15   NaN  NaN    1.0       NaN   

                                                text  author_post_karma  \
1  He even gives him self Awards to make sure he ...            17109.0   
2  RIF already announced it s shutting down at th...           118844.0   
3              I wonder which sub left on that day              50085.0   
4                  Oh I m sure this will go well lol           171705.0   
5  Oh boy i can t wait for him to say  the devs a...            70192.0   

            tag  
1  Social Media  
2  Social Media  
3  Social Media  
4  Social Media  
5  Social Media

In [14]:
comments=comments.drop(['register_index','title','url','comments'], axis=1)
comments.head()

post_id comment_id                author             datetime  score  \
1  145hyey    jnlh1vx  Illustrious_Risk3732  2023-06-09 23:42:43    1.0   
2  1453u2o    jnjtk7f       ProdigiousPlays  2023-06-09 16:47:51    1.0   
3  1453u2o    jnjtv9u                YJSubs  2023-06-09 16:49:52    1.0   
4  1453u2o    jnjusm9           jamnewton22  2023-06-09 16:55:57    1.0   
5  1453u2o    jnjvwlv            TriLink710  2023-06-09 17:03:15    1.0   

                                                text  author_post_karma  \
1  He even gives him self Awards to make sure he ...            17109.0   
2  RIF already announced it s shutting down at th...           118844.0   
3              I wonder which sub left on that day              50085.0   
4                  Oh I m sure this will go well lol           171705.0   
5  Oh boy i can t wait for him to say  the devs a...            70192.0   

            tag  
1  Social Media  
2  Social Media  
3  Social Media  
4  Social Media  
5  Social Media

In [15]:
posts = posts.drop('comment_id', axis=1)
posts.head()

register_index  post_id             author             datetime  \
125    145g5xfpost  145g5xf     lily__brunette  2023-06-09 20:40:07   
348    145f8kvpost  145f8kv  thebelsnickle1991  2023-06-09 20:03:31   
431    1447ctspost  1447cts            ionedew  2023-06-08 12:01:58   
456    145vqmzpost  145vqmz             Majnum  2023-06-10 09:31:26   
509    1466uytpost  1466uyt             777fer  2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  score  comments  \
125  https://www.reddit.com/r/technology/comments/1...  859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...    0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...    3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...   15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...   26.0      19.0   

                                                  text  author_post_karma  \
125  Reddit CEO Steve Huffman defends API changes i...              542.0   
348  Netflix s password crackdown leads to massive ...           176771.0   
431  Sam Altman says he worries making ChatGPT was ...              466.0   
456  Recent Mars Research Reveals New Discoveries a...            17593.0   
509  There s been a whopping 736 Tesla crashes invo...              105.0   

                         tag  
125             Social Media  
348                 Business  
431  Artificial Intelligence  
456                    Space  
509           Transportation

### Separating columns for comments and posts

In [16]:
cols_comment = comments.columns.to_list()
cols_comment = cols_comment[2:]
col_dict = dict.fromkeys(cols_comment)
for i in cols_comment:
    col_dict[i] = 'comment_'+i
col_dict

{'author': 'comment_author',
 'datetime': 'comment_datetime',
 'score': 'comment_score',
 'text': 'comment_text',
 'author_post_karma': 'comment_author_post_karma',
 'tag': 'comment_tag'}

In [17]:
comments.rename(columns=col_dict, inplace=True)
comments.head()

post_id comment_id        comment_author     comment_datetime  \
1  145hyey    jnlh1vx  Illustrious_Risk3732  2023-06-09 23:42:43   
2  1453u2o    jnjtk7f       ProdigiousPlays  2023-06-09 16:47:51   
3  1453u2o    jnjtv9u                YJSubs  2023-06-09 16:49:52   
4  1453u2o    jnjusm9           jamnewton22  2023-06-09 16:55:57   
5  1453u2o    jnjvwlv            TriLink710  2023-06-09 17:03:15   

   comment_score                                       comment_text  \
1            1.0  He even gives him self Awards to make sure he ...   
2            1.0  RIF already announced it s shutting down at th...   
3            1.0              I wonder which sub left on that day     
4            1.0                  Oh I m sure this will go well lol   
5            1.0  Oh boy i can t wait for him to say  the devs a...   

   comment_author_post_karma   comment_tag  
1                    17109.0  Social Media  
2                   118844.0  Social Media  
3                    50085.0  Social Media  
4                   171705.0  Social Media  
5                    70192.0  Social Media

In [18]:
posts.drop(['register_index', 'text'], axis=1, inplace=True)

In [19]:
posts.head()

post_id             author             datetime  \
125  145g5xf     lily__brunette  2023-06-09 20:40:07   
348  145f8kv  thebelsnickle1991  2023-06-09 20:03:31   
431  1447cts            ionedew  2023-06-08 12:01:58   
456  145vqmz             Majnum  2023-06-10 09:31:26   
509  1466uyt             777fer  2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  score  comments  \
125  https://www.reddit.com/r/technology/comments/1...  859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...    0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...    3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...   15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...   26.0      19.0   

     author_post_karma                      tag  
125              542.0             Social Media  
348           176771.0                 Business  
431              466.0  Artificial Intelligence  
456            17593.0                    Space  
509              105.0           Transportation

In [20]:
cols_post = ['author', 'datetime', 'score', 'tag']
col_dict = dict.fromkeys(cols_post)
for i in cols_post:
    col_dict[i] = 'post_'+i
col_dict

{'author': 'post_author',
 'datetime': 'post_datetime',
 'score': 'post_score',
 'tag': 'post_tag'}

In [21]:
posts.rename(columns=col_dict, inplace=True)
posts.head()

post_id        post_author        post_datetime  \
125  145g5xf     lily__brunette  2023-06-09 20:40:07   
348  145f8kv  thebelsnickle1991  2023-06-09 20:03:31   
431  1447cts            ionedew  2023-06-08 12:01:58   
456  145vqmz             Majnum  2023-06-10 09:31:26   
509  1466uyt             777fer  2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  post_score  comments  \
125  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...         0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...         3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...        15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...        26.0      19.0   

     author_post_karma                 post_tag  
125              542.0             Social Media  
348           176771.0                 Business  
431              466.0  Artificial Intelligence  
456            17593.0                    Space  
509              105.0           Transportation

### Merging data

In [22]:
data = posts.merge(comments, on='post_id', how='left')
print(data.shape)
data.head()

(85953, 16)


post_id     post_author        post_datetime  \
0  145g5xf  lily__brunette  2023-06-09 20:40:07   
1  145g5xf  lily__brunette  2023-06-09 20:40:07   
2  145g5xf  lily__brunette  2023-06-09 20:40:07   
3  145g5xf  lily__brunette  2023-06-09 20:40:07   
4  145g5xf  lily__brunette  2023-06-09 20:40:07   

                                               title  \
0  Reddit CEO Steve Huffman defends API changes i...   
1  Reddit CEO Steve Huffman defends API changes i...   
2  Reddit CEO Steve Huffman defends API changes i...   
3  Reddit CEO Steve Huffman defends API changes i...   
4  Reddit CEO Steve Huffman defends API changes i...   

                                                 url  post_score  comments  \
0  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
1  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
2  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
3  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
4  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   

   author_post_karma      post_tag comment_id     comment_author  \
0              542.0  Social Media    jnkvhp0         lunarmedic   
1              542.0  Social Media    jnkvmqh        CL-MotoTech   
2              542.0  Social Media    jnkti6m  whoneedsusernames   
3              542.0  Social Media    jnkyr2b         celtic1888   
4              542.0  Social Media    jnl5w5w    InternetArtisan   

      comment_datetime  comment_score  \
0  2023-06-09 21:01:30          484.0   
1  2023-06-09 21:02:30          378.0   
2  2023-06-09 20:47:55          172.0   
3  2023-06-09 21:24:12          125.0   
4  2023-06-09 22:15:35           19.0   

                                        comment_text  \
0   Direct link to AMA  https   www reddit com r ...   
1  The last I checked he had 14 responses total  ...   
2                  Incorrect title    poorly defends   
3  He ran off to his bunker after spouting nonsen...   
4   One of his most telling answers came in respo...   

   comment_author_post_karma   comment_tag  
0                    15835.0  Social Media  
1                    46444.0  Social Media  
2                    82843.0  Social Media  
3                   592646.0  Social Media  
4                   119550.0  Social Media

### Checking whether post tag is same with comment tag

In [23]:
data.loc[~data['post_tag'].eq(data['comment_tag'])]

post_id          post_author        post_datetime  \
5775   141fc0m        Stiven_Crysis  2023-06-05 13:54:38   
13987  1436x6l           Saltedline  2023-06-07 07:57:32   
15269  149dnod           whatsalamp  2023-06-14 17:25:12   
24243  145mowf             Ssider69  2023-06-10 01:18:42   
37444  14jfvaj            100GB-CSV  2023-06-26 12:22:30   
45477  14h0qny          TommyAdagio  2023-06-23 14:57:49   
49613  14g5evf               Majnum  2023-06-22 14:43:34   
53200  14fd99k               konorM  2023-06-21 17:03:02   
58978  14eagk6     OregonTripleBeam  2023-06-20 12:57:30   
74080  14b0az1      1_Portgas_D_Ace  2023-06-16 15:48:56   
82578  14a3ia3  ChickenTeriyakiBoy1  2023-06-15 14:13:16   
85440  149jhb8          ethereal3xp  2023-06-14 21:18:57   

                                                   title  \
5775   With PowerVia, Intel Achieves a Chipmaking Bre...   
13987  Supercomputer Fugaku to help develop Japanese ...   
15269  An Anti-Porn App Put Him in Jail and His Famil...   
24243               Engine treatment can slash jet noise   
37444  WSJ News Exclusive | Databricks Strikes $1.3 B...   
45477  Gearlinx emerges from stealth to deliver ‘Netw...   
49613  Satellite-based laser communication: A rival t...   
53200  Drones take to the waves: Saildrones are getti...   
58978  Portland's emergency communications bureau lau...   
74080  NASA agreements to support work on commercial ...   
82578  A Shady Chinese Firm’s Encryption Chips Got In...   
85440  Broadcom squeezed Samsung, now South Korea’s s...   

                                                     url  post_score  \
5775   https://www.reddit.com/r/technology/comments/1...        14.0   
13987  https://www.reddit.com/r/technology/comments/1...        13.0   
15269  https://www.reddit.com/r/technology/comments/1...         0.0   
24243  https://www.reddit.com/r/technology/comments/1...        36.0   
37444  https://www.reddit.com/r/technology/comments/1...        10.0   
45477  https://www.reddit.com/r/technology/comments/1...         0.0   
49613  https://www.reddit.com/r/technology/comments/1...         2.0   
53200  https://www.reddit.com/r/technology/comments/1...        32.0   
58978  https://www.reddit.com/r/technology/comments/1...        10.0   
74080  https://www.reddit.com/r/technology/comments/1...         6.0   
82578  https://www.reddit.com/r/technology/comments/1...        14.0   
85440  https://www.reddit.com/r/technology/comments/1...         7.0   

       comments  author_post_karma                 post_tag comment_id  \
5775        1.0             1357.0                 Hardware        NaN   
13987       1.0            27471.0  Artificial Intelligence        NaN   
15269       1.0             3107.0                  Privacy        NaN   
24243       1.0           322400.0           Transportation        NaN   
37444       1.0              100.0         Machine Learning        NaN   
45477       1.0             5198.0       Networking/Telecom        NaN   
49613       1.0            17654.0                 Hardware        NaN   
53200       1.0            78233.0      Robotics/Automation        NaN   
58978       1.0           257290.0  Artificial Intelligence        NaN   
74080       1.0              231.0                    Space        NaN   
82578       1.0            16752.0                 Security        NaN   
85440       1.0            32522.0       Nanotech/Materials        NaN   

      comment_author comment_datetime  comment_score comment_text  \
5775             NaN              NaN            NaN          NaN   
13987            NaN              NaN            NaN          NaN   
15269            NaN              NaN            NaN          NaN   
24243            NaN              NaN            NaN          NaN   
37444            NaN              NaN            NaN          NaN   
45477            NaN              NaN            NaN          NaN   
49613            NaN              NaN            NaN          

Above code also finds out the posts that have comments but it does not exist in the data.

In [24]:
data = data.loc[~data['comment_text'].isna()]
print(data.shape)
data.head()

(85941, 16)


post_id     post_author        post_datetime  \
0  145g5xf  lily__brunette  2023-06-09 20:40:07   
1  145g5xf  lily__brunette  2023-06-09 20:40:07   
2  145g5xf  lily__brunette  2023-06-09 20:40:07   
3  145g5xf  lily__brunette  2023-06-09 20:40:07   
4  145g5xf  lily__brunette  2023-06-09 20:40:07   

                                               title  \
0  Reddit CEO Steve Huffman defends API changes i...   
1  Reddit CEO Steve Huffman defends API changes i...   
2  Reddit CEO Steve Huffman defends API changes i...   
3  Reddit CEO Steve Huffman defends API changes i...   
4  Reddit CEO Steve Huffman defends API changes i...   

                                                 url  post_score  comments  \
0  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
1  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
2  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
3  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
4  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   

   author_post_karma      post_tag comment_id     comment_author  \
0              542.0  Social Media    jnkvhp0         lunarmedic   
1              542.0  Social Media    jnkvmqh        CL-MotoTech   
2              542.0  Social Media    jnkti6m  whoneedsusernames   
3              542.0  Social Media    jnkyr2b         celtic1888   
4              542.0  Social Media    jnl5w5w    InternetArtisan   

      comment_datetime  comment_score  \
0  2023-06-09 21:01:30          484.0   
1  2023-06-09 21:02:30          378.0   
2  2023-06-09 20:47:55          172.0   
3  2023-06-09 21:24:12          125.0   
4  2023-06-09 22:15:35           19.0   

                                        comment_text  \
0   Direct link to AMA  https   www reddit com r ...   
1  The last I checked he had 14 responses total  ...   
2                  Incorrect title    poorly defends   
3  He ran off to his bunker after spouting nonsen...   
4   One of his most telling answers came in respo...   

   comment_author_post_karma   comment_tag  
0                    15835.0  Social Media  
1                    46444.0  Social Media  
2                    82843.0  Social Media  
3                   592646.0  Social Media  
4                   119550.0  Social Media

## Finding out invalid data

In [25]:
data.loc[data['comment_author'].isin(set(data['post_author']))]

post_id      post_author        post_datetime  \
21     145g5xf   lily__brunette  2023-06-09 20:40:07   
22     145g5xf   lily__brunette  2023-06-09 20:40:07   
30     145g5xf   lily__brunette  2023-06-09 20:40:07   
51     145g5xf   lily__brunette  2023-06-09 20:40:07   
61     145g5xf   lily__brunette  2023-06-09 20:40:07   
...        ...              ...                  ...   
85806  149eob7        OneOk2189  2023-06-14 18:05:54   
85816  149egcv       spsheridan  2023-06-14 17:57:09   
85859  149drws  giuliomagnifico  2023-06-14 17:30:04   
85905  149drws  giuliomagnifico  2023-06-14 17:30:04   
85913  149drws  giuliomagnifico  2023-06-14 17:30:04   

                                                   title  \
21     Reddit CEO Steve Huffman defends API changes i...   
22     Reddit CEO Steve Huffman defends API changes i...   
30     Reddit CEO Steve Huffman defends API changes i...   
51     Reddit CEO Steve Huffman defends API changes i...   
61     Reddit CEO Steve Huffman defends API changes i...   
...                                                  ...   
85806  Paul Graham says remote work ‘does work initia...   
85816  A team of physicists recently announced that t...   
85859  Iran's 'quantum processor' turned out to be a ...   
85905  Iran's 'quantum processor' turned out to be a ...   
85913  Iran's 'quantum processor' turned out to be a ...   

                                                     url  post_score  \
21     https://www.reddit.com/r/technology/comments/1...       859.0   
22     https://www.reddit.com/r/technology/comments/1...       859.0   
30     https://www.reddit.com/r/technology/comments/1...       859.0   
51     https://www.reddit.com/r/technology/comments/1...       859.0   
61     https://www.reddit.com/r/technology/comments/1...       859.0   
...                                                  ...         ...   
85806  https://www.reddit.com/r/technology/comments/1...         0.0   
85816  https://www.reddit.com/r/technology/comments/1...        56.0   
85859  https://www.reddit.com/r/technology/comments/1...      1561.0   
85905  https://www.reddit.com/r/technology/comments/1...      1561.0   
85913  https://www.reddit.com/r/technology/comments/1...      1561.0   

       comments  author_post_karma            post_tag comment_id  \
21        232.0              542.0        Social Media    jnl4ayk   
22        232.0              542.0        Social Media    jnlawh9   
30        232.0              542.0        Social Media    jnl48nw   
51        232.0              542.0        Social Media    jnl1pfs   
61        232.0              542.0        Social Media    jnkxku3   
...         ...                ...                 ...        ...   
85806      39.0             4627.0             Society    jo5e0s5   
85816      10.0            17704.0  Nanotech/Materials    jo5hw9o   
85859     129.0            94361.0            Hardware    jo75xm0   
85905     129.0            94361.0            Hardware    jo624yz   
85913     129.0            94361.0            Hardware    jo8272u   

         comment_author     comment_datetime  comment_score  \
21              ledim35  2023-06-09 22:03:37            3.0   
22        Crack_uv_N0on  2023-06-09 22:54:07            3.0   
30              ledim35  2023-06-09 22:03:13            1.0   
51               oDDmON  2023-06-09 21:44:54           69.0   
61     LittleRickyPemba  2023-06-09 21:15:58           80.0   
...                 ...                  ...            ...   
85806        bitfriend6  2023-06-14 20:58:03            3.0   
85816         [deleted]  2023-06-14 21:26:04            3.0   
85859         [deleted]  2023-06-15 06:47:23            0.0   
85905         [deleted]  2023-06-15 00:07:16            7.0   
85913         [deleted]  2023-06-15 13:23:47            1.0   

                                            comment_text  \
21                                He s just being silly    
22     The beatings will continue until mo

# Comments Data

### Check `deleted` or `removed` comments

In [26]:
comments['comment_text']=comments['comment_text'].astype(str)
comments['comment_text']=comments['comment_text'].str.strip()

In [27]:
comments.loc[comments['comment_text'].isin(['deleted','removed'])]['comment_author'].shape

(2407,)

In [28]:
set(comments.loc[comments['comment_text'].isin(['deleted','removed'])]['comment_author'])

{'[deleted]'}

In [29]:
comments = comments.loc[~comments['comment_text'].isin(['deleted','removed'])]
print(comments.shape)
comments.head()

(83534, 8)


post_id comment_id        comment_author     comment_datetime  \
1  145hyey    jnlh1vx  Illustrious_Risk3732  2023-06-09 23:42:43   
2  1453u2o    jnjtk7f       ProdigiousPlays  2023-06-09 16:47:51   
3  1453u2o    jnjtv9u                YJSubs  2023-06-09 16:49:52   
4  1453u2o    jnjusm9           jamnewton22  2023-06-09 16:55:57   
5  1453u2o    jnjvwlv            TriLink710  2023-06-09 17:03:15   

   comment_score                                       comment_text  \
1            1.0  He even gives him self Awards to make sure he ...   
2            1.0  RIF already announced it s shutting down at th...   
3            1.0                I wonder which sub left on that day   
4            1.0                  Oh I m sure this will go well lol   
5            1.0  Oh boy i can t wait for him to say  the devs a...   

   comment_author_post_karma   comment_tag  
1                    17109.0  Social Media  
2                   118844.0  Social Media  
3                    50085.0  Social Media  
4                   171705.0  Social Media  
5                    70192.0  Social Media

### Derive hour and weekday

In [30]:
comments['comment_datetime'] = pd.to_datetime(comments['comment_datetime'])
comments

post_id comment_id        comment_author    comment_datetime  \
1      145hyey    jnlh1vx  Illustrious_Risk3732 2023-06-09 23:42:43   
2      1453u2o    jnjtk7f       ProdigiousPlays 2023-06-09 16:47:51   
3      1453u2o    jnjtv9u                YJSubs 2023-06-09 16:49:52   
4      1453u2o    jnjusm9           jamnewton22 2023-06-09 16:55:57   
5      1453u2o    jnjvwlv            TriLink710 2023-06-09 17:03:15   
...        ...        ...                   ...                 ...   
86911  149drws    jo7dx9v          talonredwing 2023-06-15 08:45:37   
86912  149drws    jo77nni             MrCamlost 2023-06-15 07:11:44   
86913  149drws    jo5i09h                a4mula 2023-06-14 21:26:52   
86914  149drws    jo7eblf                 ninta 2023-06-15 08:51:38   
86915  149drws    jo5izf2               Sorge74 2023-06-14 21:34:09   

       comment_score                                       comment_text  \
1                1.0  He even gives him self Awards to make sure he ...   
2                1.0  RIF already announced it s shutting down at th...   
3                1.0                I wonder which sub left on that day   
4                1.0                  Oh I m sure this will go well lol   
5                1.0  Oh boy i can t wait for him to say  the devs a...   
...              ...                                                ...   
86911            2.0                            You guys are all quants   
86912            0.0                 I can t make my mind up about this   
86913            2.0  I dunno  you re trying to drag me into rabbit ...   
86914            4.0  Keep trying  At some point your brainwave will...   
86915            1.0                             More like spider holes   

       comment_author_post_karma   comment_tag  
1                        17109.0  Social Media  
2                       118844.0  Social Media  
3                        50085.0  Social Media  
4                       171705.0  Social Media  
5                        70192.0  Social Media  
...                          ...           ...  
86911                     1927.0      Hardware  
86912                     2456.0      Hardware  
86913                    38276.0      Hardware  
86914                     9256.0      Hardware  
86915                   170228.0      Hardware  

[83534 rows x 8 columns]

In [31]:
comments['comment_hour'] = comments['comment_datetime'].dt.strftime('%H')
comments['comment_weekday'] = comments['comment_datetime'].dt.strftime('%a')
comments.head()

post_id comment_id        comment_author    comment_datetime  \
1  145hyey    jnlh1vx  Illustrious_Risk3732 2023-06-09 23:42:43   
2  1453u2o    jnjtk7f       ProdigiousPlays 2023-06-09 16:47:51   
3  1453u2o    jnjtv9u                YJSubs 2023-06-09 16:49:52   
4  1453u2o    jnjusm9           jamnewton22 2023-06-09 16:55:57   
5  1453u2o    jnjvwlv            TriLink710 2023-06-09 17:03:15   

   comment_score                                       comment_text  \
1            1.0  He even gives him self Awards to make sure he ...   
2            1.0  RIF already announced it s shutting down at th...   
3            1.0                I wonder which sub left on that day   
4            1.0                  Oh I m sure this will go well lol   
5            1.0  Oh boy i can t wait for him to say  the devs a...   

   comment_author_post_karma   comment_tag comment_hour comment_weekday  
1                    17109.0  Social Media           23             Fri  
2                   118844.0  Social Media           16             Fri  
3                    50085.0  Social Media           16             Fri  
4                   171705.0  Social Media           16             Fri  
5                    70192.0  Social Media           17             Fri

### Handling null values

In [32]:
comments.isna().sum()

post_id                         0
comment_id                      0
comment_author                  0
comment_datetime                0
comment_score                   0
comment_text                    0
comment_author_post_karma    1613
comment_tag                     0
comment_hour                    0
comment_weekday                 0
dtype: int64

In [33]:
comments.loc[comments['comment_author_post_karma'].isna()]

post_id comment_id   comment_author    comment_datetime  comment_score  \
414    1442xur    jnegb54  Educationalsrsr 2023-06-08 14:51:40            2.0   
482    145hyey    jnloq9h        [deleted] 2023-06-10 00:45:12            7.0   
872    144gbnn    jnh0rqh       FloodMoose 2023-06-09 01:13:28            4.0   
873    144gbnn    jnj6lfd       FloodMoose 2023-06-09 14:19:37            1.0   
1034   1496yl6    jo4blcr        [deleted] 2023-06-14 16:31:13           -1.0   
...        ...        ...              ...                 ...            ...   
86596  149h82b    joh0v9s        [deleted] 2023-06-17 14:17:15            1.0   
86646  149h82b    joaqawp         ulogomaa 2023-06-16 01:24:58            1.0   
86691  149h82b    jo6jzm4         Andre5k5 2023-06-15 02:41:49            3.0   
86724  149h82b    joaq40f         ulogomaa 2023-06-16 01:23:19            1.0   
86822  149drws    jo75xm0        [deleted] 2023-06-15 06:47:23            0.0   

                                            comment_text  \
414    They target low income minorities   particular...   
482    Well I will answer it  With the delete account...   
872    I hope that whatever it is  it uses decentrali...   
873    Like the twatter    where it turns into a for ...   
1034   Pushing a narrative  means framing events in a...   
...                                                  ...   
86596  Can someone explain this topic to me  I ve rea...   
86646              upvoted  everything you said is right   
86691  Remember when the founder dildod himself to ow...   
86724  sorry you cant handle opinions being exchanged...   
86822  They are speed running turning their country i...   

       comment_author_post_karma         comment_tag comment_hour  \
414                          NaN  Networking/Telecom           14   
482                          NaN        Social Media           00   
872                          NaN        Social Media           01   
873                          NaN        Social Media           14   
1034                         NaN        Social Media           16   
...                          ...                 ...          ...   
86596                        NaN        Social Media           14   
86646                        NaN        Social Media           01   
86691                        NaN        Social Media           02   
86724                        NaN        Social Media           01   
86822                        NaN            Hardware           06   

      comment_weekday  
414               Thu  
482               Sat  
872               Fri  
873               Fri  
1034              Wed  
...               ...  
86596             Sat  
86646             Fri  
86691             Thu  
86724             Fri  
86822             Thu  

[1613 rows x 10 columns]

Fill null values for author karma with zero.

In [34]:
comments.fillna(0, inplace=True)

In [35]:
comments.groupby('comment_tag').count().sort_values('comment_text', ascending=False)

post_id  comment_id  comment_author  \
comment_tag                                                    
Social Media               34853       34853           34853   
Business                   13918       13918           13918   
Transportation              5140        5140            5140   
Hardware                    4448        4448            4448   
Security                    3634        3634            3634   
Society                     3288        3288            3288   
Artificial Intelligence     3226        3226            3226   
Energy                      2974        2974            2974   
Networking/Telecom          2576        2576            2576   
Politics                    2124        2124            2124   
Biotechnology               1395        1395            1395   
Software                    1331        1331            1331   
Privacy                     1294        1294            1294   
Space                       1020        1020            1020   
ADBLOCK WARNING             1020        1020            1020   
Nanotech/Materials           783         783             783   
Crypto                       227         227             227   
Robotics/Automation          118         118             118   
Net Neutrality                77          77              77   
TechSupport                   52          52              52   
Machine Learning              36          36              36   

                         comment_datetime  comment_score  comment_text  \
comment_tag                                                              
Social Media                        34853          34853         34853   
Business                            13918          13918         13918   
Transportation                       5140           5140          5140   
Hardware                             4448           4448          4448   
Security                             3634           3634          3634   
Society                              3288           3288          3288   
Artificial Intelligence              3226           3226          3226   
Energy                               2974           2974          2974   
Networking/Telecom                   2576           2576          2576   
Politics                             2124           2124          2124   
Biotechnology                        1395           1395          1395   
Software                             1331           1331          1331   
Privacy                              1294           1294          1294   
Space                                1020           1020          1020   
ADBLOCK WARNING                      1020           1020          1020   
Nanotech/Materials                    783            783           783   
Crypto                                227            227           227   
Robotics/Automation                   118            118           118   
Net Neutrality                         77             77            77   
TechSupport                            52             52            52   
Machine Learning                       36             36            36   

                         comment_author_post_karma  comment_hour  \
comment_tag                                                        
Social Media                                 34853         34853   
Business                                     13918         13918   
Transportation                                5140          5140   
Hardware                                      4448          4448   
Security                                      3634          3634   
Society                                       3288          3288   
Artificial Intelligence                       3226          3226   
Energy                                        2974          2974   
Networking/Telecom                            2576          2576   
Politics                                      2124          2124   
Biotechnology                    

## Posts Data

In [37]:
posts.isna().sum()

post_id               0
post_author           0
post_datetime         0
title                 0
url                   0
post_score            0
comments              0
author_post_karma    12
post_tag              0
dtype: int64

In [38]:
# Handling null values
posts.fillna(0, inplace=True)

### Derive hour and weekday

In [39]:
posts['post_datetime'] = pd.to_datetime(posts['post_datetime'])
posts.head()

post_id        post_author       post_datetime  \
125  145g5xf     lily__brunette 2023-06-09 20:40:07   
348  145f8kv  thebelsnickle1991 2023-06-09 20:03:31   
431  1447cts            ionedew 2023-06-08 12:01:58   
456  145vqmz             Majnum 2023-06-10 09:31:26   
509  1466uyt             777fer 2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  post_score  comments  \
125  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...         0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...         3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...        15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...        26.0      19.0   

     author_post_karma                 post_tag  
125              542.0             Social Media  
348           176771.0                 Business  
431              466.0  Artificial Intelligence  
456            17593.0                    Space  
509              105.0           Transportation

In [40]:
posts['post_hour'] = posts['post_datetime'].dt.strftime('%H')
posts['post_weekday'] = posts['post_datetime'].dt.strftime('%a')
posts.head()

post_id        post_author       post_datetime  \
125  145g5xf     lily__brunette 2023-06-09 20:40:07   
348  145f8kv  thebelsnickle1991 2023-06-09 20:03:31   
431  1447cts            ionedew 2023-06-08 12:01:58   
456  145vqmz             Majnum 2023-06-10 09:31:26   
509  1466uyt             777fer 2023-06-10 18:05:28   

                                                 title  \
125  Reddit CEO Steve Huffman defends API changes i...   
348  Netflix's password crackdown leads to massive ...   
431  Sam Altman says he worries making ChatGPT was ...   
456  Recent Mars Research Reveals New Discoveries a...   
509  There's been a whopping 736 Tesla crashes invo...   

                                                   url  post_score  comments  \
125  https://www.reddit.com/r/technology/comments/1...       859.0     232.0   
348  https://www.reddit.com/r/technology/comments/1...         0.0      14.0   
431  https://www.reddit.com/r/technology/comments/1...         3.0      10.0   
456  https://www.reddit.com/r/technology/comments/1...        15.0       2.0   
509  https://www.reddit.com/r/technology/comments/1...        26.0      19.0   

     author_post_karma                 post_tag post_hour post_weekday  
125              542.0             Social Media        20          Fri  
348           176771.0                 Business        20          Fri  
431              466.0  Artificial Intelligence        12          Thu  
456            17593.0                    Space        09          Sat  
509              105.0           Transportation        18          Sat

### Adding post and comment time difference

In [41]:
comments = comments.merge(posts[['post_id','post_datetime', 'post_hour','post_weekday']], on='post_id', how='left')
comments.head()

post_id comment_id        comment_author    comment_datetime  \
0  145hyey    jnlh1vx  Illustrious_Risk3732 2023-06-09 23:42:43   
1  1453u2o    jnjtk7f       ProdigiousPlays 2023-06-09 16:47:51   
2  1453u2o    jnjtv9u                YJSubs 2023-06-09 16:49:52   
3  1453u2o    jnjusm9           jamnewton22 2023-06-09 16:55:57   
4  1453u2o    jnjvwlv            TriLink710 2023-06-09 17:03:15   

   comment_score                                       comment_text  \
0            1.0  He even gives him self Awards to make sure he ...   
1            1.0  RIF already announced it s shutting down at th...   
2            1.0                I wonder which sub left on that day   
3            1.0                  Oh I m sure this will go well lol   
4            1.0  Oh boy i can t wait for him to say  the devs a...   

   comment_author_post_karma   comment_tag comment_hour comment_weekday  \
0                    17109.0  Social Media           23             Fri   
1                   118844.0  Social Media           16             Fri   
2                    50085.0  Social Media           16             Fri   
3                   171705.0  Social Media           16             Fri   
4                    70192.0  Social Media           17             Fri   

        post_datetime post_hour post_weekday  
0 2023-06-09 21:51:15        21          Fri  
1 2023-06-09 12:25:11        12          Fri  
2 2023-06-09 12:25:11        12          Fri  
3 2023-06-09 12:25:11        12          Fri  
4 2023-06-09 12:25:11        12          Fri

In [42]:
comments.isna().sum()

post_id                      0
comment_id                   0
comment_author               0
comment_datetime             0
comment_score                0
comment_text                 0
comment_author_post_karma    0
comment_tag                  0
comment_hour                 0
comment_weekday              0
post_datetime                0
post_hour                    0
post_weekday                 0
dtype: int64

In [43]:
comments['time_diff'] = (comments['comment_datetime']-comments['post_datetime']).dt.total_seconds()/3600

comments

post_id comment_id        comment_author    comment_datetime  \
0      145hyey    jnlh1vx  Illustrious_Risk3732 2023-06-09 23:42:43   
1      1453u2o    jnjtk7f       ProdigiousPlays 2023-06-09 16:47:51   
2      1453u2o    jnjtv9u                YJSubs 2023-06-09 16:49:52   
3      1453u2o    jnjusm9           jamnewton22 2023-06-09 16:55:57   
4      1453u2o    jnjvwlv            TriLink710 2023-06-09 17:03:15   
...        ...        ...                   ...                 ...   
83529  149drws    jo7dx9v          talonredwing 2023-06-15 08:45:37   
83530  149drws    jo77nni             MrCamlost 2023-06-15 07:11:44   
83531  149drws    jo5i09h                a4mula 2023-06-14 21:26:52   
83532  149drws    jo7eblf                 ninta 2023-06-15 08:51:38   
83533  149drws    jo5izf2               Sorge74 2023-06-14 21:34:09   

       comment_score                                       comment_text  \
0                1.0  He even gives him self Awards to make sure he ...   
1                1.0  RIF already announced it s shutting down at th...   
2                1.0                I wonder which sub left on that day   
3                1.0                  Oh I m sure this will go well lol   
4                1.0  Oh boy i can t wait for him to say  the devs a...   
...              ...                                                ...   
83529            2.0                            You guys are all quants   
83530            0.0                 I can t make my mind up about this   
83531            2.0  I dunno  you re trying to drag me into rabbit ...   
83532            4.0  Keep trying  At some point your brainwave will...   
83533            1.0                             More like spider holes   

       comment_author_post_karma   comment_tag comment_hour comment_weekday  \
0                        17109.0  Social Media           23             Fri   
1                       118844.0  Social Media           16             Fri   
2                        50085.0  Social Media           16             Fri   
3                       171705.0  Social Media           16             Fri   
4                        70192.0  Social Media           17             Fri   
...                          ...           ...          ...             ...   
83529                     1927.0      Hardware           08             Thu   
83530                     2456.0      Hardware           07             Thu   
83531                    38276.0      Hardware           21             Wed   
83532                     9256.0      Hardware           08             Thu   
83533                   170228.0      Hardware           21             Wed   

            post_datetime post_hour post_weekday  time_diff  
0     2023-06-09 21:51:15        21          Fri   1.857778  
1     2023-06-09 12:25:11        12          Fri   4.377778  
2     2023-06-09 12:25:11        12          Fri   4.411389  
3     2023-06-09 12:25:11        12          Fri   4.512778  
4     2023-06-09 12:25:11        12          Fri   4.634444  
...                   ...       ...          ...        ...  
83529 2023-06-14 17:30:04        17          Wed  15.259167  
83530 2023-06-14 17:30:04        17          Wed  13.694444  
83531 2023-06-14 17:30:04        17          Wed   3.946667  
83532 2023-06-14 17:30:04        17          Wed  15.359444  
83533 2023-06-14 17:30:04        17          Wed   4.068056  

[83534 rows x 14 columns]

In [44]:
comments.to_csv('comments.csv', index=False)
print('comments.csv created')

comments.csv created
